### PR TITLE
New tests + code simplification

### PR DESCRIPTION
**Pull request by ryantmer**
- Use hidden attributes with backing computed properties rather than dom-if
- Move .json()-ing outside of handlers into Promise chain to reduce nesting
- Other nesting-reduction changes for fun
- Add tests

TODO:

- [x] merge #45 and #47 and rebase this

### CODE REVIEW

File: src/d2l-image-banner-overlay.html [hunk 1]
```diff
@@ -160,12 +174,17 @@ <h1 id="courseName">[[_courseName]]</h1>
...
 			_getOrganizationInfo: function() {
-				window.d2lfetch
+				return window.d2lfetch
 					.fetch(new Request(this.organizationUrl, {
 						headers: new Headers({
 							Accept: 'application/vnd.siren+json'
 						})
 					}))
+					.then(function(response) {
+						if (response.ok) {
```
Comment (CodeBaboon) [inline src/d2l-image-banner-overlay.html]: What happens when `!response.ok`? Feels like we'll pass non-json badness into `_onOrganizationResponse` which will explode attempting to parse it.

File: src/d2l-image-banner-overlay.html [hunk 2]
```diff
@@ -176,32 +195,32 @@ <h1 id="courseName">[[_courseName]]</h1>
 			},
 			_onSetCourseImage: function(e) {
 				var org = this._parseSiren(e.detail.organization);
-				if (!org.getLinkByRel) { return; }
+				var orgSelfLink = (org.getLinkByRel('self') || {}).href || '';
 
-				var orgSelfLink = (org.getLinkByRel('self') || {}).href || '',
-					urlPosition = orgSelfLink.indexOf(this.organizationUrl);
+				if (orgSelfLink.toLowerCase() !== this.organizationUrl.toLowerCase()) {
```
Comment (CodeBaboon) [inline src/d2l-image-banner-overlay.html]: Is this accurate? It no longer matches the logic from before the change, which didn't need them to match completely but rather for the `orgSelfLink` to just end with the `organizationurl` (and also used to make sure the case matched).

File: src/d2l-image-banner-overlay.html (hunk 1 above)
Comment (ryantmer) [reply src/d2l-image-banner-overlay.html]: Ah, good call yeah. `_onOrganizationResponse` was previously just returning in that case, but I've lost that here. I'll restore that behaviour (should probably reject and show an error image or something, but...)

Comment (ryantmer) [reply src/d2l-image-banner-overlay.html]: (added in 3165493)

**Comment by ryantmer** [comment]
Tested the following things:

- Menu items are properly controlled by permissions; users without the correct permission don't see the item; users with neither permission don't see the menu at all
- Clicking to remove the banner removes the banner and presents the `d2l-alert`
- Clicking "undo" on the alert restores the banner
- Clicking "X" on the alert closes the alert
- Clicking to change the image opens the image selector overlay
- Selecting a new image updates the banner image
- If a My Courses widget is on the same homepage, https://github.com/Brightspace/d2l-my-courses-ui/pull/435 results in the image for the current course also updating in My Courses (not yet merged, but it's a distinct change that I just happened to notice)
- If a My Courses widget is on the same homepage and the course image is changed via the tile, the banner also updates if it's the same course (doesn't update if you change a different course's image in My Courses)
- Everything is tab-navigable as before

@CodeBaboon any additional test cases that you can think of?

**Comment by CodeBaboon** [comment]
@ryantmer nope that seems like it tests all the things!

**Pull request merged**